src/spaces/openstreetmap.jl

```diff
@@ -0,0 +1,591 @@
+export OpenStreetMapSpace, OSMSpace, TEST_MAP
+export osm_random_road_position,
+    osm_plan_route,
+    osm_map_coordinates,
+    osm_road_length,
+    osm_is_stationary,
+    osm_random_route!
+
+struct OpenStreetMapSpace <: DiscreteSpace
+    m::OpenStreetMapX.MapData
+    edges::Vector{CartesianIndex{2}}
+    s::Vector{Vector{Int}}
+end
+
+"""
+    OpenStreetMapSpace(path::AbstractString; kwargs...)
+Create a space residing on the Open Street Map (OSM) file provided via `path`.
+
+The abbreviation `OSMSpace` may be used interchangeably.
+
+Much of the functionality of this space is provided by interfacing with
+[OpenStreetMapX.jl](https://github.com/pszufe/OpenStreetMapX.jl), for example the two
+keyword arguments `use_cache = false` and `trim_to_connected_graph = true` can be
+passed into the [`OpenStreetMapX.get_map_data`](@ref) function.
+
+For details on how to obtain an OSM file for your use case, consult the OpenStreetMapX.jl
+README. We provide a variable `TEST_MAP` to use as a `path` for testing.
+
+This space represents the underlying map as a *continuous* entity choosing accuracy over
+performance. If your solution can tolerate routes to and from intersections only, a
+faster implementation can be achieved by using the
+[graph representation](https://pszufe.github.io/OpenStreetMapX.jl/stable/reference/#OpenStreetMapX.MapData)
+of your map provided by OpenStreetMapX.jl. For tips on how to implement this, see our
+integration example: [Social networks with LightGraphs.jl](@ref).
+
+## The OSMAgent
+
+The base properties for an agent residing on an `OSMSpace` are as follows:
+```julia
+mutable struct OSMAgent <: AbstractAgent
+    id::Int
+    pos::Tuple{Int,Int,Float64}
+    route::Vector{Int}
+    destination::Tuple{Int,Int,Float64}
+end
+```
+
+Current `pos`ition and `destination` tuples are represented as
+`(start intersection index, finish intersection index, distance travelled in meters)`.
+The `route` is an ordered list of intersections, providing a path to reach `destination`.
+
+Further details can be found in [`OSMAgent`](@ref).
+
+## Routing
+
+There are three ways to generate a route, depending on the situation.
+1. Using [`add_agent!`](@ref)`(start, finish, model)`, which will calculate a route automatically for a new agent.
```

Comment:
That's a bit confusing for me. You add an agent *and* plan its route? If the agent has extra properties, how are they created? is it with the syntax of extra keywords? 

Perhaps this should be a different function or split into two functions? one that adds the route to the agent? Because if an agent already has a valid route we can use `add_agent_pos!` instead, instead of adding further dispatches to `add_agent!`. It might be a bit complicated if `add_agent!` does too many things.

Reply:
Saw the docstring of add_agent now. Given that `plan_route` also accepts additional keywords, I'd vote to not overload `add_agent!` here for clarity.

Reply:
It's just the same as any add_agent! method. But sure, I've dropped that method and edited the example.